---
title: "liv_go"
author: "Dani Blumstein"
date: "2023-12-02"
output: html_document
---

```{r}
gene_ton_fun <- function(res_object,verson_num,ensembl)
  {
res_object$ensembl <- sapply( strsplit( rownames(res_object), split="\\+" ), "[", 1 )

#querying with the ENSEMBL gene id and requesting the Entrez gene id and HGNC gene symbol.
library("biomaRt")
genemap <- getBM( attributes = c("ensembl_gene_id", "hgnc_symbol"),
                  filters = "ensembl_gene_id",
                  values = rownames(res_object),
                  mart = ensembl)

idx <- match(rownames(res_object), genemap$ensembl_gene_id )
res_object$hgnc_symbol <- genemap$hgnc_symbol[ idx ]
hgnc_symbol <- res_object$hgnc_symbol

mcols(liv_dds_trt) <- cbind(mcols(liv_dds_trt), hgnc_symbol)
res_object$hgnc_symbol <- rowData(liv_dds_trt)$hgnc_symbol

liv_dds_trt <- estimateSizeFactors(liv_dds_trt)

FC <- 0.6
p <- 50e-3

keyvals <- rep('grey75', nrow(res_object))
names(keyvals) <- rep('NS', nrow(res_object))

keyvals[which(abs(res_object$log2FoldChange) > FC & res_object$pvalue > p)] <- 'grey50'
names(keyvals)[which(abs(res_object$log2FoldChange) > FC & res_object$pvalue > p)] <- 'log2FoldChange'

keyvals[which(abs(res_object$log2FoldChange) < FC & res_object$pvalue < p)] <- 'grey25'
names(keyvals)[which(abs(res_object$log2FoldChange)  < FC & res_object$pvalue < p)] <- '-Log10Q'

keyvals[which(res_object$log2FoldChange < -FC & res_object$pvalue < p)] <- 'blue2'
names(keyvals)[which(res_object$log2FoldChange  < -FC & res_object$pvalue < p)] <- 'Signif. down-regulated'

keyvals[which(res_object$log2FoldChange > FC & res_object$pvalue < p)] <- 'red2'
names(keyvals)[which(res_object$log2FoldChange > FC & res_object$pvalue < p)] <- 'Signif. up-regulated'

liv_volcano2 <- EnhancedVolcano(res_object,
                lab = res_object$hgnc_symbol,
                x = 'log2FoldChange',
                y = 'pvalue',
                #selectLab = rownames(res_object)[which(names(keyvals) %in% c('NS','log2FoldChange','-Log10Q','low','high'))],
                xlim = c(-6.5,6.5),
                xlab = bquote(~Log[2]~ 'fold change'),
                ylab = bquote(~-Log[10] ~ italic(P)),
                title = 'yes vs no',
                pCutoff = 50e-3,
                FCcutoff = 0.6,
                pointSize = 2.5,
                labSize = 4.5,
                #shape = c(6, 4, 2, 11, 15),
                #colCustom = keyvals,
                colAlpha = 0.75,
                legendPosition = 'right',
                legendLabSize = 15,
                legendIconSize = 5.0,
                drawConnectors = FALSE,
                widthConnectors = 0.5,
                colConnectors = 'grey50',
                gridlines.major = TRUE,
                gridlines.minor = FALSE,
                border = 'partial',
                borderWidth = 1.5,
                borderColour = 'black')
ggsave(paste0("figures/liv/liv_volcano2_", verson_num, ".png"), width = 8, height = 8, dpi = 600)
liv_volcano2

return(res_object)
}

gene_ton_fig <- function(res_object,verson_num, ensembl)
  {
  
anno_df <- data.frame(gene_id = res_object$ensembl, gene_name = res_object$hgnc_symbol)
vst_macrophage <- varianceStabilizingTransformation(liv_dds_trt)

bg_ids <- rownames(liv_dds_trt)[rowSums(counts(liv_dds_trt)) > 0]
bg_symbols <- mapIds(org.Hs.eg.db,
                     keys = bg_ids,
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")
 
# res_enrich object
#library("AnnotationDbi")
de_symbols_IFNg_vs_naive <- deseqresult2df(res_object, FDR =0.05)$hgnc_symbol

# res_enrich object ------------------------------------------------------------
 topgoDE_macrophage_IFNg_vs_naive <-
  pcaExplorer::topGOtable(de_symbols_IFNg_vs_naive,
                          bg_symbols,
                          ontology = "BP",
                          mapping = "org.Hs.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

res_enrich <- shake_topGOtableResult(topgoDE_macrophage_IFNg_vs_naive)
res_enrich <- get_aggrscores(res_enrich = res_enrich,
                                        res_de = res_object,
                                        annotation_obj = anno_df,
                                        aggrfun = mean)

enhance_table <- enhance_table(res_enrich,
                   res_object,
                   n_gs = 30,
                   annotation_obj = anno_df,
                   chars_limit = 60)
enhance_table
ggsave(paste0("GO/liv/liv_enhance_table_", verson_num, ".png"), width = 8, height =15, dpi = 600)


gs_alluvial(res_enrich = res_enrich,
            res_de = res_object,
            annotation_obj = anno_df,
            n_gs = 30)

volcano_plot<- gs_volcano(res_enrich,
           p_threshold = 0.05,
           color_by = "aggr_score",
           volcano_labels = 10,
           gs_ids = NULL,
           plot_title = "my volcano")
volcano_plot
ggsave(paste0("GO/liv/liv_volcano_plot_", verson_num, ".png"), width = 8, height = 4.6, dpi = 600)


mds <- gs_mds(res_enrich,
       res_object,
       anno_df,
       n_gs = 200,
       gs_ids = NULL,
       similarity_measure = "kappa_matrix",
       mds_k = 2,
       mds_labels = 5,
       mds_colorby = "z_score",
       gs_labels = NULL,
       plot_title = NULL) 
mds
ggsave(paste0("GO/liv/liv_mds_", verson_num, ".png"), width = 8, height = 4.6, dpi = 600)


gs_dendro(res_enrich,
          n_gs = 30,
          gs_dist_type = "kappa", 
          clust_method = "ward.D2",
          color_leaves_by = "z_score",
          size_leaves_by = "gs_pvalue",
          color_branches_by = "clusters",
          create_plot = TRUE)

ggs <- ggs_graph(res_enrich,
                 res_de = res_object,
                 anno_df,
                 n_gs = 30)
ggs

scores_mat <- gs_scores(
  se = vst_macrophage,
  res_de = res_object,
  res_enrich = res_enrich,
  annotation_obj = anno_df
)
gs_scoresheat(scores_mat,n_gs = 30)


#geneset_heatmap1 <- gs_heatmap(se = vst_macrophage,
#           res_de = res_object,
#           res_enrich = res_enrich,
#           annotation_obj = anno_df,
#           geneset_id = "GO:0097746",
#           cluster_columns = TRUE,
#           anno_col_info = c("sex", "trt")
#)
#geneset_heatmap1

#geneset_volcano <- signature_volcano(res_de = liv_res_trt,
#                  res_enrich = res_enrich,
#                  annotation_obj = anno_df,
#                  geneset_id = "GO:0048514",
#                  FDR = 0.05,
#                  color = "#1a81c2"
#)
#ggsave("GO/liv_geneset_volcano.png", width = 8, height = 4.6, dpi = 600)
#geneset_volcano
}
```

```{r liv data}
liv_countdata <- dplyr::select(countdata, contains("liv"))
liv_samples <-samples %>% filter(tissue == "liv") 
liv_samples$trt <- as.factor(liv_samples$trt)
liv_samples$sex <- as.factor(liv_samples$sex)

liv_dds_trt <- DESeqDataSetFromMatrix(countData = liv_countdata,
                                       colData = liv_samples,
                                       design = ~ trt + sex)


#filter out genes where there are less than 8 samples with normalized counts greater than or equal to 10.
liv_dds_trt <- estimateSizeFactors(liv_dds_trt)
idx <- rowSums(counts(liv_dds_trt, normalized=TRUE) >= 10 ) >= 8

liv_dds_trt <- liv_dds_trt[idx,]
print(paste("Number of genes after filtering by total expression:", nrow(liv_dds_trt)))
matrix <- as.data.frame(counts(liv_dds_trt))

#liv_dds_trt$trt_combo <- factor(paste0(liv_dds_trt$trt, liv_dds_trt$sex))
#design(liv_dds_trt) <- ~ pro_weight + trt_combo

liv_dds_trt <- DESeq(liv_dds_trt,test="Wald")
#res but with ensid name
liv_res_trt <- results(liv_dds_trt, alpha = 0.05)
liv_res_trt$symbol <- mcols(liv_res_trt)$symbol

# how many are "significant"?
table(liv_res_trt[,"padj"] < 0.05)
summary(liv_res_trt)


#BiocManager::install("vsn")
#library("vsn")
vsd <- varianceStabilizingTransformation(liv_dds_trt, blind = FALSE)
meanSdPlot(assay(vsd))


pcaData <- plotPCA(vsd, intgroup = c("trt", "sex"), returnData = TRUE)
# get percent variation
percentVar <- round(100 * attr(pcaData, "percentVar"))
# pca code
pca <- ggplot(pcaData, aes(x = PC1, y = PC2, color = trt, shape = sex, name=name)) +
  stat_ellipse(aes(group = trt, linetype = trt), type = "t", level = 0.95, size = 1.25, show.legend = FALSE) +
  #scale_linetype_manual(values=c("twodash", "longdash", "solid"), guide = FALSE) +
  geom_point(size = 3, show.legend = TRUE) + 
  #scale_color_manual(values = c("orange1", "red1", "yellow2", "chartreuse4")) +
  #scale_shape_manual(values = c(8,9,15,16,10,17,18,11,13)) + 
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() + theme_bw() +
  guides(shape = guide_legend(order = 1),color = guide_legend(order = 2)) 
ggplotly(pca)
ggsave("figures/liv_PCA.png", width = 8, height = 4.6, dpi = 600)


# crea sample distances and sample distance matrix
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
heatmapcolors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix, clusr_rows = FALSE, clusr_cols = FALSE, col = heatmapcolors)
```

```{r}
#BiocManager::install("fastmatrix")
#BiocManager::install("pcaExplorer")
library(pcaExplorer)
annos$ensembl <-as.character(annos$ensid)

#rename transcript rows to ensid names
rownames(liv_dds_trt)=annos$ensid[match(names(liv_dds_trt),annos$gene)]
#remove the NAs (those transcripts dont have a ensid name)
liv_dds_trt <- liv_dds_trt[complete.cases(rownames(liv_dds_trt)), ]
liv_dds_trt <- liv_dds_trt[!rownames(liv_dds_trt) == "NULL", ] 

liv_dds_trt <- estimateSizeFactors(liv_dds_trt)
normalized_counts <- counts(liv_dds_trt, normalized=TRUE)
#NOTE: DESeq2 doesnâ€™t actually use normalized counts, rather it uses the raw counts and models the normalization inside the Generalized Linear Model (GLM). These normalized counts will be useful for downstream visualization of results, but cannot be used as input to DESeq2 or any other tools that peform differential expression analysis which use the negative binomial model.
```

### Differential expression analyses

I will run some differential expression analyses now. I want to account for sex, trt, and proportional weight loss in my analyses.

Run differential expression analysis. What is the effect of sex?

```{r}
ddssex <- DESeqDataSetFromMatrix(countData = liv_countdata,
                                       colData = liv_samples,
                                       design = ~ trt + sex)

#filter out genes where there are less than 8 samples with normalized counts greater than or equal to 10.
ddssex <- estimateSizeFactors(ddssex)
idx <- rowSums(counts(ddssex, normalized=TRUE) >= 10 ) >= 8

ddssex <- ddssex[idx,]
print(paste("Number of genes after filtering by total expression:", nrow(ddssex)))
matrix <- as.data.frame(counts(ddssex))


ddssex <- DESeq(ddssex, test="Wald")
ressex <- results(ddssex, alpha = 0.05, contrast = c("sex", "M", "F"))
ressex$transcript <- mcols(ressex)$transcript
liv_res_trt$symbol <- mcols(liv_res_trt)$symbol

# how many are "significant"?
table(ressex[,"padj"] < 0.05)
summary(ressex)
```

best model for downstream things right now
```{r}
ddstrt <- DESeqDataSetFromMatrix(countData = liv_countdata,
                                       colData = liv_samples,
                                       design = ~ trt + sex)


#filter out genes where there are less than 8 samples with normalized counts greater than or equal to 10.
ddstrt <- estimateSizeFactors(ddstrt)
idx <- rowSums(counts(ddstrt, normalized=TRUE) >= 10 ) >= 8

ddstrt <- ddstrt[idx,]
print(paste("Number of genes after filtering by total expression:", nrow(ddstrt)))
matrix <- as.data.frame(counts(ddstrt))

ddstrt <- DESeq(ddstrt, test="Wald")

#rename transcript rows to ensid names
rownames(ddstrt)=annos$ensid[match(names(ddstrt),annos$gene)]
#remove the NAs (those transcripts dont have a ensid name)
ddstrt <- ddstrt[complete.cases(rownames(ddstrt)), ]
ddstrt <- ddstrt[!rownames(ddstrt) == "NULL", ] 

ddstrt <- estimateSizeFactors(ddstrt)
normalized_counts <- counts(ddstrt, normalized=TRUE)

restrt <- results(ddstrt, alpha = 0.05, contrast = c("trt", "standard", "lowfat"))
restrt$transcript <- mcols(restrt)$transcript
restrt$symbol <- mcols(restrt)$symbol

# how many are "significant"?
table(restrt[,"padj"] < 0.05)
summary(restrt)

restrt_liv <-restrt


vsd <- varianceStabilizingTransformation(ddstrt, blind = FALSE)
meanSdPlot(assay(vsd))


pcaData <- plotPCA(vsd, intgroup = c("trt", "sex"), returnData = TRUE)
# get percent variation
percentVar <- round(100 * attr(pcaData, "percentVar"))
# pca code
pca <- ggplot(pcaData, aes(x = PC1, y = PC2, color = trt, shape = sex, name=name)) +
  stat_ellipse(aes(group = trt, linetype = trt), type = "t", level = 0.95, size = 1.25, show.legend = FALSE) +
  #scale_linetype_manual(values=c("twodash", "longdash", "solid"), guide = FALSE) +
  geom_point(size = 3, show.legend = TRUE) + 
  #scale_color_manual(values = c("orange1", "red1", "yellow2", "chartreuse4")) +
  #scale_shape_manual(values = c(8,9,15,16,10,17,18,11,13)) + 
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() + theme_bw() +
  guides(shape = guide_legend(order = 1),color = guide_legend(order = 2)) 
ggplotly(pca)
```


```{r}
restrt <- gene_ton_fun(restrt,0,ensembl)
write.csv(restrt, file="results/GOresults/liv/liv_de_GO.csv")
#gene_ton_fig(restrt,0,ensembl)

anno_df <- data.frame(gene_id = restrt$ensembl, gene_name = restrt$hgnc_symbol)
vst_macrophage <- varianceStabilizingTransformation(liv_dds_trt)

bg_ids <- rownames(liv_dds_trt)[rowSums(counts(liv_dds_trt)) > 0]
bg_symbols <- mapIds(org.Hs.eg.db,
                     keys = bg_ids,
                     column = "SYMBOL",
                     keytype = "ENSEMBL",
                     multiVals = "first")
 
# res_enrich object
#library("AnnotationDbi")
de_symbols_IFNg_vs_naive <- deseqresult2df(restrt, FDR =0.05)$hgnc_symbol

# res_enrich object ------------------------------------------------------------
 topgoDE_macrophage_IFNg_vs_naive <-
  pcaExplorer::topGOtable(de_symbols_IFNg_vs_naive,
                          bg_symbols,
                          ontology = "BP",
                          mapping = "org.Hs.eg.db",
                          geneID = "symbol",
                          topTablerows = 30)

res_enrich <- shake_topGOtableResult(topgoDE_macrophage_IFNg_vs_naive)
res_enrich_liv <- get_aggrscores(res_enrich = res_enrich,
                                        res_de = restrt,
                                        annotation_obj = anno_df,
                                        aggrfun = mean)

#GeneTonic(dds = liv_dds_trt,
#          res_de = restrt,
#          res_enrich = res_enrich_liv,
#          annotation_obj = anno_df,
#          project_id = "myFirstGeneTonic")


gene_list <- c("ENSG00000229314","ENSG00000141485","ENSG00000146678","ENSG00000142973","ENSG00000168209","ENSG00000151014")
plist <- c()
for (i in gene_list) {
  
plist[[i]] <- gene_plot(liv_dds_trt,
          gene = c(i),
          intgroup = "trt",
          assay = "counts",
          annotation_obj = anno_df,
          plot_type = "boxplot",
          labels_display = F)

}
liv_DE_plots <- gridExtra::grid.arrange(grobs = plist, ncol=1)
ggsave("figures/liv/liv_DE_plots.png", plot = liv_DE_plots, width = 8, height = 20)
```

```{r}
up_liv <- filter(as.data.frame(restrt), log2FoldChange > 0, padj < 0.05)
gostres_liv_up <- gost(query = up_liv$hgnc_symbol, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)

go_liv_up <- as.data.frame(cbind(gostres_liv_up$result$term_id,gostres_liv_up$result$term_name, gostres_liv_up$result$term_size, gostres_liv_up$result$p_value))
go_liv_up$tissue <- "liv"
go_liv_up$DE <- "up"
go_liv_up <- dplyr::filter(go_liv_up, grepl('GO', V1))
go_liv_up <- top_n(go_liv_up, 20, V4)

down_liv <- filter(as.data.frame(restrt), log2FoldChange < 0, padj < 0.05)
gostres_liv_down <- gost(query = down_liv$hgnc_symbol, 
                organism = "hsapiens", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = FALSE, 
                user_threshold = 0.05, correction_method = "g_SCS", 
                domain_scope = "annotated", custom_bg = NULL, 
                numeric_ns = "", sources = NULL, as_short_link = FALSE, highlight = TRUE)

go_liv_down <- as.data.frame(cbind(gostres_liv_down$result$term_id,gostres_liv_down$result$term_name, gostres_liv_down$result$term_size, gostres_liv_down$result$p_value))
go_liv_down$tissue <- "liv"
go_liv_down$DE <- "down"
go_liv_down <- dplyr::filter(go_liv_down, grepl('GO', V1))
go_liv_down <- top_n(go_liv_down, 20, V4)

go_liv <- list(go_liv_down,go_liv_up)
go_liv <- go_liv %>% reduce(full_join, by=c("V1","V2"))
```
