---
title: "RNAseq"
author: "Dani Blumsin"
da: "7/8/2022"
output: html_document
---

knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

devtools::install_github('kevinblighe/EnhancedVolcano')
BiocManager::install("DESeq2")
BiocManager::install("apeglm")
BiocManager::install("topGO")
BiocManager::install("BiocParallel")
library("remotes")
remotes::install_github("federicomarini/GeneTonic", 
                        dependencies = TRUE, build_vignettes = TRUE)
                        
```{r}
setwd("/Users/danielleblumstein/Documents/UNH/rnaseq/diet_final")
#call adams fancy functions
source("adams_fancy_functions.R")
ensembl = useMart( "ensembl", dataset = "hsapiens_gene_ensembl" )
```

```{r}
library("GeneTonic")
library("macrophage")
library("org.Hs.eg.db")
library("AnnotationDbi")
library(DESeq2)
library(dplyr)
library(foreach)
library(data.table)
library(splines)
library(ggthemes)
library(scales)
library(gridExtra)
library(tidyr)
library(pheatmap)
library(RColorBrewer)
library(ggplot2)
library(apeglm)
library(tidyverse)
library(topGO)
library(GO.db)
library(readr)
library(readxl)
library(lubridate)
library(patchwork)
library(EnhancedVolcano)
dir.create("figures")
dir.create("figures/TalkGeneFigures")
dir.create("figures4publication")
dir.create("results")
dir.create("results/DEresults")
dir.create("results/GOresults")
dir.create("results/WGCNAresults")
```


make sample name spreadsheet
```{r}
### List all samples from expression data 
# get the directory/path for each sample in this study
base_dir <- getwd()

# get directories
filename <- list.files(path = "data/", pattern = "gene.counts", full.names = FALSE, recursive = FALSE)

# append full file path to each sample id
files <- file.path(base_dir, "data/", filename) # files = directory + salmon directory + sample name + quantifictaion file name
names(files) <- "data/"
all(file.exists(files)) # do these all actually exist?

# Make sample spreadsheet 
sample <- as.data.frame(filename)

# get sample
sample$individal <- filename  %>% gsub(pattern = "gene.counts", replacement = "")

###############come back to##############
# get tissue
sample$tissue <- unlist(lapply(strsplit(sample$individal, split = "_"),"[[", 3))

# get sex
sample$sex <- unlist(lapply(strsplit(sample$individal, split = "_"),"[[", 2))

# get trt
sample$trt <- unlist(lapply(strsplit(sample$individal, split = "_"),"[[", 4))

# get sample ID
sample$Animal_ID <- unlist(lapply(strsplit(sample$individal, split = "_"),"[[", 1))

#grabbing the phys data and electro/ weight data. currently comming from diff script but it happens to be in my global enviro. will need to change this 

#cages <- read.csv("~/Documents/UNH/metabolic chamber/diet/cages") 
#electrolyte_data <- read_excel("data/electrolyte_data.xlsx", na = "NA")
#col_order <- c("Animal_ID", "StartTime", "VO2", "VCO2", "H2Omg","EE_kJH", "RQ", "weight", "seconds")
#phys_data <- cages[, col_order]
#electrolyte_data$Animal_ID <- as.character(electrolyte_data$mouse_ID)
#
#
##splitting out only the phys data between 11:00 and 12:00
#noon <- period_to_seconds(hms("11:00:00")):period_to_seconds(hms("12:00:00"))
#phys_data$TimeOfDay <- ifelse(phys_data$seconds %in% noon, "noon", "other")
#phys_data <- phys_data[phys_data$TimeOfDay == "noon", ]
#
#mean_VO2<- setNames(aggregate(phys_data$VO2, list(phys_data$Animal_ID), FUN=mean),c("Animal_ID", "mean_VO2"))
#
#mean_VCO2<- setNames(aggregate(phys_data$VCO2, list(phys_data$Animal_ID), FUN=mean),c("Animal_ID", "mean_VCO2"))
#
#mean_EE<- setNames(aggregate(phys_data$EE_kJH, list(phys_data$Animal_ID), FUN=mean),c("Animal_ID", "mean_EE"))
#
#mean_RQ<- setNames(aggregate(phys_data$RQ, list(phys_data$Animal_ID), FUN=mean),c("Animal_ID", "mean_RQ"))
#
#mean_H2Omg<- setNames(aggregate(phys_data$H2Omg, list(phys_data$Animal_ID), FUN=mean),c("Animal_ID", "mean_H2Omg"))
#
##merge all data frames together
#samples_phys<- Reduce(function(x, y) merge(x, y, all=TRUE), list(mean_VO2, mean_VCO2, mean_RQ, mean_H2Omg, mean_EE,sample))
#
#samples_phys$Animal_ID <- as.character(samples_phys$Animal_ID)
#
#samples <- electrolyte_data %>% right_join(samples_phys)
#
#
##samples <- samples[complete.cases(samples),]
#samples$...6 <- NULL
#samples$cage_ID <- NULL
#samples$experiment <- NULL
#samples$experiment_date <- NULL
#samples$cage_ID <- NULL
#samples$cage_number<- NULL
#samples$weight2<- NULL
#samples$weight3<- NULL
#samples$end_weight<- NULL
#samples$diet<- NULL
#samples$mouse_ID<- NULL

#############do this next!##############
#correct for multiple hypothesis for all comparisons
#make a priori hypothesis as specific as i can for all the comparisons?
samples <- sample
samples<- samples[!is.na(samples$filename),]

write.csv(samples, "results/samplespreadsheet.csv", row.names = FALSE)
#write.csv(cages, "results/cages.csv", row.names = FALSE)
```


```{r}
## Import annotation documents 
#annos <- read_delim("data/annos.txt", delim = "\t", 
#+     escape_double = FALSE, col_names = FALSE, 
#+     locale = locale(), trim_ws = TRUE)
annos <- read_csv("/Users/danielleblumstein/Documents/UNH/rnaseq/diet_final/data/annos.txt")

library("org.Hs.eg.db")
annos$ensid = mapIds(org.Hs.eg.db,
                    keys=annos$gene, 
                    column="ENSEMBL",
                    keytype="SYMBOL",
                    multiVals="first")

#######################come back to ##########################
# make spreadsheet for deseq
#filenames <- samples[,21]
```


```{r}
countdata <- data.frame(annos$transcript)
colnames(countdata) <- "transcript"
for (counts in (1:length(samples$filename)))
  {
  # import data for that sample
  toimport <- samples$filename[counts]
  tmpdata <- read.table(paste0("data/", toimport), header = FALSE, sep = "\t")
  # sample id
  samplename <- toimport  %>% gsub(pattern = "gene.counts", replacement = "")
  samplename <- gsub("_S.*", "", samplename)
  colnames(tmpdata) <- c("transcript", samplename)
  countdata <- dplyr::left_join(countdata, tmpdata, by = "transcript")
}
dim(countdata)
countdata <- data.frame(countdata, row.names = 1)
names(countdata)=str_sub(names(countdata),2)

write.csv(countdata, "countdata.csv", row.names = FALSE)
```


Quick data control. Change any "NA" values to "0"
```{r data QC}
# change all NA to 0
print(paste0("Number of NA after import:", sum(is.na(countdata))))
countdata[is.na(countdata)] <-  0
print(paste0("Number of NA after NA removal:", sum(is.na(countdata))))
print(paste0("Number of genes after import:", nrow(countdata)))

## remove all unknown proteins
countdata <- countdata[!rownames(countdata) == "proteins of unknown function", ]
print(paste("Number of genes after removing all 'proteins of unknown function':", nrow(countdata)))

dim(countdata)
```


The count data table is currently a dataframe of counts, with each row representing a transcript in the genome. Some genes are represented by multiple transcripts in the genome, and hence our count data table. Here I will manipulate the data so that we can do a gene level analysis.

```{R Create a data frame of gene level count data}
# make rownames column 1
countdata <- setDT(countdata, keep.rownames = "transcript")[]

# add gene symbol
countdata2 <- dplyr::left_join(annos, countdata, by = "transcript")
dim(countdata2)

# change all NA to 0
print(paste("Number of NA after import:", sum(is.na(countdata2))))

countdata2 <- countdata2 %>% 
    mutate_at(5:215, ~replace_na(.,0))

# combine all counts that map to the same gene
countdata3 <- aggregate(countdata2[, 5:215], list(countdata2$gene), sum)
dim(countdata3)

countdata3[is.na(countdata3)] <-  0
colnames(countdata3)[1] <- "Gene"
countdata <- countdata3

# make gene name column
countdata <- data.frame(countdata, row.names = 1)
print(paste("Number of genes after combining all counts that map to the same gene:", nrow(countdata)))

for ( col in 1:ncol(countdata)){
    colnames(countdata)[col] <-  sub("X*", "", colnames(countdata)[col])
}

# save gene-level count data
write.table(countdata, "data/gene.level.count.data.tsv", row.names = TRUE, sep = "\t")
```


### DESeq data creation 

```{r DESEQ creation}
dim(countdata)
dim(samples)
# crea DESeq dataset
countdata <- countdata+1
dds <- DESeqDataSetFromMatrix(countData = countdata,
                                       colData = samples,
                                       design = ~ sex + tissue + trt)
```


### Crea a PCA

```{R PCA}
vsd <- varianceStabilizingTransformation(dds, blind = FALSE)
# get PC1 and PC2 data
pcaData <- plotPCA(vsd, intgroup = c("tissue","trt"), returnData = TRUE)
# get percent variation
percentVar <- round(100 * attr(pcaData, "percentVar"))
# pca code
pca <- ggplot(pcaData, aes(x = PC1, y = PC2, color = tissue, shape = trt, name=name)) +
  stat_ellipse(aes(group = tissue, linetype = tissue), type = "t", level = 0.95, size = 1.25, show.legend = FALSE) +
  #scale_linetype_manual(values=c("twodash", "longdash", "solid"), guide = FALSE) +
  geom_point(size = 3, show.legend = TRUE) + 
  #scale_color_manual(values = c("orange1", "red1", "yellow2", "chartreuse4")) +
  #scale_shape_manual(values = c(8,9,15,16,10,17,18,11,13)) + 
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() + theme_bw() +
  guides(shape = guide_legend(order = 1),color = guide_legend(order = 2)) 
# save
ggsave("figures/PCA.png", width = 8, height = 4.6, dpi = 600)


library(plotly)
ggplotly(pca)

p <- ggplotly(pca)
```


### Make the final PCA for the publication

```{R PCA for publication}
# make sure levels of morphs are correctly specified for colors
# pca code
ggplot(pcaData, aes(x = PC1, y = PC2, color = tissue)) +
  stat_ellipse(aes(group = tissue, linetype = tissue), type = "t", level = 0.95, size = 1.25, show.legend = FALSE) +
  #scale_linetype_manual(values=c("twodash", "longdash", "solid"), guide = FALSE) +
  geom_point(size = 3, show.legend = TRUE) + 
  #scale_color_manual(values = c("orange1", "red1", "yellow2", "chartreuse4")) +
  #scale_shape_manual(values = c(8,9,15,16,10,17,18,11,13)) + 
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() + theme_bw() +
  guides(shape = guide_legend(order = 1),color = guide_legend(order = 2)) 
  #annota("xt", label = "Ranitomeya fantastica", x = 45, y = -22.5, size = 4, colour = "black", fontface = "italic") +
  #annota("xt", label = "Ranitomeya imitator", x = 38, y = 22, size = 4, colour = "black", fontface = "italic") +
  #annota("xt", label = "Ranitomeya variabilis", x = -45, y = 12, size = 4, colour = "black", fontface = "italic") 
  
ggsave("figures4publication/PCA.png", width = 8, height = 4.6, dpi = 600)
```


### Crea a heatmap

Make a heatmap with rows/columns ordered by populations then time?
#doesn't do much for my data yet

```{r HEATWAVE}
# crea sample distances and sample distance matrix
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
# rename columns/rows so they are inrpretable
#rownames(sampleDistMatrix) <- gsub(patrn = "_M.*$", replacement = "", rownames(sampleDistMatrix))
#colnames(sampleDistMatrix) <- gsub(patrn = "_M.*$", replacement = "", colnames(sampleDistMatrix))
# plot heatmap
heatmapcolors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix, clusr_rows = FALSE, clusr_cols = FALSE, col = heatmapcolors)
# plot and save heatmap
png("figures/heatmap.png", width = 14.63, height = 10, units = "in", res = 600)
pheatmap(sampleDistMatrix, clusr_rows = FALSE, clusr_cols = FALSE, col = heatmapcolors)
```

### Differential expression analyses

I will run some differential expression analyses now. I want to account for sex, trt, and tissue in my analyses.


Run differential expression analysis. What is the effect of sex?

```{r LRT sex}
ddssex <- DESeqDataSetFromMatrix(countData = countdata,
                                       colData = samples,
                                       design = ~ trt + tissue + sex)
# run LRT test
ddssex <- DESeq(ddssex, test="LRT", reduced = ~ trt + tissue)
ressex <- results(ddssex)
ressex$transcript <- mcols(ressex)$transcript
# how many are "significant"?
table(ressex[,"padj"] < 0.05)
```

```{r LRT trt}
ddstrt <- DESeqDataSetFromMatrix(countData = countdata,
                                       colData = samples,
                                       design = ~ sex + tissue + trt)
# run LRT test
ddstrt <- DESeq(ddstrt, test="LRT", reduced = ~ sex + tissue)
restrt <- results(ddstrt,)
restrt$symbol <- mcols(restrt)$symbol
# how many are "significant"?
table(restrt[,"padj"] < 0.05)
```

```{r LRT tissue}
ddstissue <- DESeqDataSetFromMatrix(countData = countdata,
                                       colData = samples,
                                       design = ~ trt + sex + tissue)
# run LRT test
ddstissue <- DESeq(ddstissue, test="LRT", reduced = ~ trt + sex)
restissue <- results(ddstissue)
restissue$symbol <- mcols(restissue)$symbol
# how many are "significant"?
table(restissue[,"padj"] < 0.05)
```

*look for genes expressed in one and not the other can do this in edgeR*

need a prioi for this function to work
```{r DE genes by sex, eval = FALSE}
#SigGeneWrapper(ressex, 0.05, "sex")
```

```{r DE genes by trt, eval = FALSE}
#SigGeneWrapper(restrt, 0.05, "trt")
```

```{r DE genes by tissue, eval = FALSE}
#SigGeneWrapper(restissue, 0.05, "tissue")
```



Now I will run these test for each tissue independently. 

